### PR TITLE
Java: Make more finegrained dataflow dispatch viable callable heuristic.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we modify the dataflow dispatch viable callable implementation. The main purpose of the PR is that we don't want to dispath to a synthetic/summary callable that only has generated summaries in case a source code implementation is available.

The gains are
- Reduces combinatorial blowup for dataflow analysis as the viable callable fanout is reduced when analyzing a project that has generated models.
- Improves model generation precision as generated models for a given library are not used to produce more models for the same library when models are re-generated. The change in this PR at least makes summary model generation observably idempotent for `jdk-17` and `commons-io`.
- Implementation is more aligned with the C# implementation.
- Found a bad generated `logging` summary for `getParameters`. We should consider to make a follow up and make more accurate summaries for `getParameters` and `setParameters`.

As a consequence we also need to modify some of the models (promote some `df-generated` models to `df-manual` and override some `generated` models by summary `neutrals`). The section below based on the DCA execution describes the
reasoning behind the model changes.
It should be noted that it is not strictly needed for other projects than those including parts of the `java-sdk` source code, so I don't think we need to strive for complete accuracy. 

DCA:
We see a significant speedup on the `kotlin` and `open-jdk` projects (15-20%). This is because we reduce the dataflow call dispatch fanout when both source code and a summary is available for a callable.

With respect to alert discrepancies for `open-jdk` - this is not completely surprising as it contains the `jdk` source code. It turns out that the over approximation in the model generator provides some "reasonable" summary models that are not derived by strict data flow analysis, which is why we need to promote these models.

The following was done to reduce the alert discrepancy on `open-jdk`:

- The summary models for `jave.nio.Charbuffer`, `java.nio.ByteBuffer`, `java.net.URI` are needed for `java/tainted-arithmetic` and `java/uncontrolled-arithmetic`. Furthermore, we also need to add a neutral for the method `getEncoding` in `java.util.logging.Handler` as this looks incorrect and triggers an alert.
- `java/improper-validation-of-array-index` and `java/improper-validation-of-array-construction` uses the “bad” `LogRecord.getParameters` method from `java.util.logging` on some paths - adding a `neutral` for this one (This is an example of Pojo that won’t work with the current version of the model generator)
- `java/zipslip` requires the summary for `URI.getPath` from `java.net`.
- `java/sensitive-logging` needs the promoted models from `java.security.cert` and `javax.management`.
- `java/polynomial-redos` needs the promoted models for `getMessage` in `java.util.logging`.

Investigation of the remaining alert discrepancies on `open-jdk`:
Removals:
- `java/hardcoded-credential-api-call` result is removed because the heuristic callback summary is not being used as the source code is available. We should probably accept this. Alternatively we should re-classify the heuristic callback summary as manual heuristic.
- `java/weak-cryptographic-algorithm` requires promoting at least three groups of models to manual models. It seems unreasonable to spend further time on this.
- `java/improper-validation-of-array-index` and `java/improper-validation-of-array-construction` used the summary for `LogRecord.getParameters` method from `java.util.logging`, which has been replaced by a neutral (so the discrepancy is expected). So alert removal is expected.
- `java/sensitive-log`: I sampled around half of the "missing" results and they all used `getParameters` (which we introduced a neutral for). So alert removal is expected.
- `java/tainted-arithmetic`: One of the removed alerts is due to the introduction of the the neutral for `getEncoding`. The other removal, I don't have a good explanation for and same goes for the two removals in `java/uncontrolled-arithmetic`.

Additions
- `java/sensitive-log` and `java/path-inection`: These alerts are not using any steps that didn't exist before. Maybe they are found now as the search space has been reduced.

For `elasticsearch` two more `java/sensitive-log` results have been found. There doesn't appear to be any reason, why they shouldn't have been found prior to this change.

### CODE REVIEW

File: java/ql/lib/ext/java.nio.model.yml [hunk 1]
```diff
@@ -5,7 +5,20 @@ extensions:
     data:
       - ["java.nio", "ByteBuffer", False, "array", "()", "", "Argument[this]", "ReturnValue", "taint", "manual"]
       - ["java.nio", "ByteBuffer", False, "get", "", "", "Argument[this]", "ReturnValue", "taint", "manual"]
-      - ["java.nio", "ByteBuffer", False, "wrap", "(byte[])", "", "Argument[0]", "ReturnValue", "taint", "manual"]
+      - ["java.nio", "ByteBuffer", True, "put", "(ByteBuffer)", "", "Argument[this]", "ReturnValue", "value", "df-manual"]
```
Comment (aschackmull) [inline java/ql/lib/ext/java.nio.model.yml:8]: Is this overload missing an `Argument[0]` to `Argument[this]` taint step?

File: java/ql/lib/ext/java.nio.model.yml [hunk 2]
```diff
@@ -5,7 +5,20 @@ extensions:
...
+      - ["java.nio", "ByteBuffer", True, "put", "(byte)", "", "Argument[this]", "ReturnValue", "value", "df-manual"]
+      - ["java.nio", "ByteBuffer", True, "put", "(byte[])", "", "Argument[0]", "Argument[this]", "taint", "df-manual"]
+      - ["java.nio", "ByteBuffer", True, "put", "(byte[])", "", "Argument[0]", "ReturnValue", "taint", "df-manual"]
+      - ["java.nio", "ByteBuffer", True, "put", "(byte[])", "", "Argument[this]", "ReturnValue", "taint", "df-manual"]
```
Comment (aschackmull) [inline java/ql/lib/ext/java.nio.model.yml:12]: For fluent taint-adding apis like this, we need just two models - not three. The third will be inferred. Also the arg zero to return should be value - not taint.
```suggestion
      - ["java.nio", "ByteBuffer", True, "put", "(byte[])", "", "Argument[0]", "Argument[this]", "taint", "df-manual"]
      - ["java.nio", "ByteBuffer", True, "put", "(byte[])", "", "Argument[0]", "ReturnValue", "value", "df-manual"]
```

File: java/ql/lib/ext/java.util.logging.model.yml [hunk 3]
```diff
@@ -46,8 +46,15 @@ extensions:
...
       pack: codeql/java-all
       extensible: neutralModel
     data:
+      - ["java.util.logging", "Handler", "getEncoding", "()", "summary", "manual"]
       - ["java.util.logging", "Logger", "isLoggable", "(Level)", "summary", "manual"]
+      - ["java.util.logging", "LogRecord", "getResourceBundle", "()", "summary", "df-manual"]
+      # If needed, a pair of manual summary models using synthetics can be made for the two
+      # neutrals below.
+      - ["java.util.logging", "LogRecord", "getParameters", "()", "summary", "manual"]
+      - ["java.util.logging", "LogRecord", "setParameters", "", "summary", "df-manual"]
```
Comment (aschackmull) [inline java/ql/lib/ext/java.util.logging.model.yml:60]: If we've by now manually verified that this is just a pojo, then I do think it's relevant to add such manual summary models using synthetic fields. Adding these as manual neutrals just seems plain wrong.

Comment (michaelnebel) [reply java/ql/lib/ext/java.util.logging.model.yml:60]: There was already a manual neutral for one of the methods - so this was to make this "consistent".
I agree - it is listed in the description as a follow up - I didn't want to also introduce this here as it was complicated enough as it is to investigate the alert discrepancies.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll [hunk 4]
```diff
@@ -40,11 +40,29 @@ private module DispatchImpl {
...
+      not (
+        // Only use summarized callables with generated summaries in case
+        // we are not able to dispatch to a source declaration.
+        // Note that if applyGeneratedModel holds it implies that there doesn't
+        // exist a manual (exact) model.
+        exists(Callable callable | callable = sourceDispatch(call) |
+          callable.fromSource() and not callable.isStub()
+        ) and
+        result.asSummarizedCallable().applyGeneratedModel()
+      ) and
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:63]: I'm slightly concerned whether this risks FNs. I.e. if some api point is intended to be overridden then you may have library implementations from which a good model was generated and also source implementations that are very different. I'm not exactly sure how to best iterate on this heuristic, but I think we should require that the static call target is `fromSource` in order to boost our confidence in the source dispatch.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll [hunk 5]
```diff
@@ -40,11 +40,29 @@ private module DispatchImpl {
     else any()
   }
 
-  /** Gets a viable implementation of the target of the given `Call`. */
+  /**
+   * Gets a viable implementation of the target of the given `Call`.
+   * The following heuristic is applied for finding the appropriate callable:
+   * 1. If an exact manual model exists, only dispatch to the summarized callable.
+   * 2. If a (non exact) manual model exists and/or if the source code is available, dispatch to both/either.
+   * 3. Only dispatch to a summarized callable (based on a generated model) if neither of the above apply.
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:48]: This is not entirely correct, but let's get an agreed semantics settled before tweaking this comment.

File: java/ql/lib/ext/java.util.logging.model.yml (hunk 3 above)
Comment (aschackmull) [reply java/ql/lib/ext/java.util.logging.model.yml:60]: Ok.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll (hunk 4 above)
Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:63]: To make sure I understand your concern.
Are you worried about the following:
Assume we have a library *L* containing
```
class Base { ... }
class A : Base { ... }
```
and we generate models based on *L*, where the implementation of `A` gives us "good" models for `Base`.

If we furthermore assume that we want to analyse some source code, which imports `L`, but also contains a "bad" implementation of `Base` (that doesn't exhibit the same flow as `A`)
```
class B : Base { ... }
```

Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:63]: Yes, exactly. It might very well be that `A` is the default implementation when interacting with the library through `Base` and that `B` is degenerate in some way - e.g. a test stub.
Another condition to potentially take into account is whether the static call target is `B` or `Base`. If the static call target is `B` then it's probably fine to disregard any generated summaries.

Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:63]: Fair point, but at least the DCA projects didn't exhibit this problematic pattern.
What do you think about changing the generated models exclusion logic to something like:
```
not (
        // Only use summarized callables with generated summaries in case
        // we are not able to dispatch to a source declaration.
        // Note that if applyGeneratedModel holds it implies that there doesn't
        // exist a manual (exact) model.
        exists(Callable staticTarget | staticTarget = call.getCallee() |
          staticTarget.fromSource() and not staticTarget.isStub()
        ) and
        exists(Callable callable | callable = sourceDispatch(call) |
          callable.fromSource() and not callable.isStub()
        ) and
        result.asSummarizedCallable().applyGeneratedModel()
      )
```
There might also be something special for constructors that needs to be takes into account - but I can try and write some tests for this.

**Comment by michaelnebel** [comment]
@aschackmull :
The condition for when we disregard generated summaries have been tightened and a couple of extra tests have been added. Summary model generation is still observably idempotent for the `jdk17` and `commons-io`.
Do you think the tightened condition looks more acceptable?

If this is acceptable, I will
(1) Run DCA again (which hopefully shows approximately the same discrepancies as before).
(2) If the DCA run from (1) looks acceptable, I will also improve the `df-manual` models and introduce models for `getParameters` and `setParameters` as a part of this PR (as you recommended) and run DCA once again.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll (hunk 4 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll]: Yes, I think that looks much better.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll [hunk 6]
```diff
@@ -40,11 +40,29 @@ private module DispatchImpl {
     else any()
   }
 
-  /** Gets a viable implementation of the target of the given `Call`. */
+  /**
+   * Gets a viable implementation of the target of the given `Call`.
+   * The following heuristic is applied for finding the appropriate callable:
+   * 1. If an exact manual model exists, only dispatch to the summarized callable.
+   * 2. If a (non exact) manual model exists and/or if the source code is available, dispatch to both/either.
+   * 3. Only dispatch to a summarized callable in case the static call target in not in source.
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:48]: I've tried to reword this to be more accurate.
```suggestion
   * The following heuristic is applied for finding the appropriate callable:
   * In general, dispatch to both any existing model and any viable source dispatch.
   * However, if the model is generated and the static call target is in the source then
   * we trust the source more than the model and skip dispatch to the model.
   * Vice versa, if the model is manual and the source dispatch has a comparatively low
   * confidence then we only dispatch to the model. Additionally, manual models that
   * match a source dispatch exactly take precedence over the source.
```

Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll:48]: I see now that I forgot to write "generated" summary in (3).
Your suggestion looks good - we should use that!

**Review by aschackmull — APPROVED** [review]
LGTM!

**Comment by michaelnebel** [comment]
DCA still looks good!

**Comment by michaelnebel** [comment]
@aschackmull : I updated some models and added some extra tests since your last review. DCA didn't show anything new/surprising.

**Review by aschackmull — APPROVED** [review]
LGTM
